src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.cxx

```diff
@@ -903,6 +923,73 @@ namespace xclcpuemhal2 {
     return size;
   }
 
+  int CpuemShim::xclRegRW(bool rd, uint32_t cu_index, uint32_t offset, uint32_t *datap)
+  {
+    if (mLogStream.is_open()) {
+      mLogStream << __func__ << ", " << std::this_thread::get_id() << ", " << "CU Idx : " << cu_index << " Offset : " << offset << " Datap : " << (*datap) << std::endl;
+    }
+    // get sorted cu addresses to match up with cu_index
+    const auto& cuidx2addr = mCoreDevice->get_cus();
+    //mLogStream << __func__ << " , cuidx2addr :  " << cuidx2addr << std::endl;
+    if (cu_index >= cuidx2addr.size()) {
+      std::string strMsg = "ERROR: [SW-EMU 20] xclRegRW - invalid CU index: " + std::to_string(cu_index);
+      mLogStream << __func__ << strMsg << std::endl;
+      return -EINVAL;
+    }
+
+    uint64_t cuAddRange = 64 * 1024;
+    for (auto cuInfo : mCURangeMap) {
+      std::string instName = cuInfo.first;
+      int cuIdx = static_cast<int>(cu_index);
+      int tmpCuIdx = xclIPName2Index(instName.c_str());
+      mLogStream << __func__ << " , instName :  " << instName  << " cuIdx : " << cuIdx << " tmpCuIdx: " << tmpCuIdx << std::endl;
+      if (tmpCuIdx == cuIdx) {
+        cuAddRange = cuInfo.second;
+        mLogStream << __func__ << " , cuAddRange :  " << cuAddRange << std::endl;
+      }
+    
+    }
+
+    if (offset >= cuAddRange || (offset & (sizeof(uint32_t) - 1)) != 0) {
+      std::string strMsg = "ERROR: [SW-EMU 21] xclRegRW - invalid CU offset: " + std::to_string(offset);
+      mLogStream << __func__ << strMsg << std::endl;
+      return -EINVAL;
+    }
+
```

Comment:
I think this function should be refactor into smaller pieces that do the various checks one-by-one.  This would make the code more readable.  No function should be more than 10+ lines in general, of course there are exceptions, but then very careful commenting would be needed.